gen/collision_test.go

```diff
@@ -0,0 +1,205 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package gen
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"testing"
+
+	tc "go.uber.org/thriftrw/gen/testdata/collision"
+	"go.uber.org/thriftrw/ptr"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/thriftrw/protocol"
+	"go.uber.org/thriftrw/wire"
+)
+
+func TestStruct(t *testing.T) {
+	tests := []struct {
+		desc string
+		x    thriftType
+	}{
+		{
+			"StructCollision",
+			&tc.StructCollision{
+				CollisionField:  true,
+				CollisionField2: "true",
+			},
+		},
+		{
+			"StructCollision2 (struct_collision)",
+			&tc.StructCollision2{
+				CollisionField:  true,
+				CollisionField2: "such trueness",
+			},
+		},
+		{
+			"PrimitiveContainers",
+			&tc.PrimitiveContainers{
+				A: []string{"arbre", "fleur"},
+				B: map[string]struct{}{
+					"croissant": struct{}{},
+					"baguette":  struct{}{},
+				},
+				C: map[string]string{
+					"voiture": "bleue",
+					"cammion": "rouge",
+				},
+			},
+		},
+		{
+			"StructConstant (struct_constant)",
+			tc.StructConstant,
+		},
+		{
+			"UnionCollision .CollisionField",
+			&tc.UnionCollision{
+				CollisionField: ptr.Bool(true),
+			},
+		},
+		{
+			"UnionCollision .collision_field",
+			&tc.UnionCollision{
+				CollisionField2: ptr.String("so true"),
+			},
+		},
+		{
+			"UnionCollision2 (union_collision) .CollisionField",
+			&tc.UnionCollision2{
+				CollisionField: &[]bool{true}[0],
+			},
+		},
+		{
+			"UnionCollision (union_collision) .collision_field",
+			&tc.UnionCollision2{
+				CollisionField2: ptr.String("true indeed"),
+			},
+		},
+	}
+	for _, tt := range tests {
+		roundTrip(t, tt.x, tt.desc)
+	}
+}
+
+func TestConstant(t *testing.T) {
+	require.Equal(t, tc.StructCollision2{
+		CollisionField2: "false indeed",
+	}, *tc.StructConstant)
+}
+
+func TestWithDefault(t *testing.T) {
+	a := &tc.WithDefault{
+		Pouet: &tc.StructCollision2{
+			CollisionField2: "false indeed",
+		},
+	}
+	b := &tc.WithDefault{}
+
+	a = troughWire(t, a, "WithDefault{filled in}").(*tc.WithDefault)
+	b = troughWire(t, b, "WithDefault{}").(*tc.WithDefault)
+	require.Equal(t, a, b)
+}
+
+func TestMyEnum(t *testing.T) {
+	tests := []struct {
+		e tc.MyEnum
+		n string
+		v int64
+	}{
+		{tc.MyEnumX, "X", 123},
+	}
+	for _, tt := range tests {
+		assert.Equal(t, tt.n, tt.e.String())
+		assert.Equal(t, tt.v, int64(tt.e))
+	}
+}
+
+func TestMyEnum2(t *testing.T) {
+	tests := []struct {
+		e tc.MyEnum2
+		n string
+		v int64
+	}{
+		{tc.MyEnum2X, "X", 12},
+	}
+	for _, tt := range tests {
+		assert.Equal(t, tt.n, tt.e.String())
+		assert.Equal(t, tt.v, int64(tt.e))
+	}
+}
+
+func troughWire(t *testing.T, x thriftType, msg string, args ...interface{}) thriftType {
```

Comment:
yeah roundTrip is used already in another file. Also, yes, I meant through. I never know which one is which.